src/libraries/System.Private.CoreLib/src/System/IO/Path.Unix.cs

```diff
@@ -34,7 +34,7 @@ public static string GetFullPath(string path)
 
             // We would ideally use realpath to do this, but it resolves symlinks, requires that the file actually exist,
             // and turns it into a full path, which we only want if fullCheck is true.
-            string collapsedString = PathInternal.RemoveRelativeSegments(path, PathInternal.GetRootLength(path));
+            string collapsedString = Path.RemoveRedundantSegments(path.AsSpan());
```

Comment:
Is this going to regress performance of the existing GetFullPath API? This API was not allocating anything when there was nothing to normalize. It is going to allocate a new string after this change.

Reply:
This is how the old API used to look. There was one allocation that was always being done, when calling the constructor:

```cs
        internal static string RemoveRelativeSegments(string path, int rootLength)	
        {	
            var sb = new ValueStringBuilder(stackalloc char[260 /* PathInternal.MaxShortPath */]);	

            if (RemoveRelativeSegments(path.AsSpan(), rootLength, ref sb))	
            {	
                path = sb.ToString();	
            }	

            sb.Dispose();	
            return path;	
        }
```

Then the overload that takes 3 arguments, which is being called in the `if` clause, would do the job of copying all the characters from the original path into the `ValueStringBuilder` instance.

The allocations were being done before. The main difference is that I added extra logic to handle some edge cases that the former method was not handling.

Reply:
I ran a unit test calling `Path.GetFullPath`, with my changes, inside an infinite while loop, and collected time spent using `dotnet trace` for a few seconds, then viewed the results in speedscope.app.

On the right side you can see the stack of calls, and the length of each bar is the time spent on each method. Most of the time is spent in CPU_TIME inside the `TryRemoveRedundantSegments` method.

![image](https://user-images.githubusercontent.com/1175054/100666166-6d85aa80-330d-11eb-8cc2-5631b71afb7a.png)

Reply:
> There was one allocation that was always being done, when calling the constructor:

There is allocation done on GC heap in the current implementation of GetFullPath API when there is nothing to normalize.